src/api/java/com/minecolonies/api/colony/requestsystem/StandardFactoryController.java

```diff
@@ -1,45 +1,65 @@
 package com.minecolonies.api.colony.requestsystem;
 
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.reflect.TypeToken;
+import com.minecolonies.api.colony.requestsystem.factory.FactoryVoidInput;
 import com.minecolonies.api.colony.requestsystem.factory.IFactory;
 import com.minecolonies.api.colony.requestsystem.factory.IFactoryController;
 import com.minecolonies.api.util.Log;
+import com.minecolonies.api.util.ReflectionUtils;
 import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.util.Tuple;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 import java.util.HashMap;
-import java.util.Map;
-
-import static com.minecolonies.api.util.constant.Suppression.RAWTYPES;
-import static com.minecolonies.api.util.constant.Suppression.UNCHECKED;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
 
 /**
  * Default implementation of a FactoryController
  * Singleton.
  */
-public final class StandardFactoryController implements IFactoryController
+public class StandardFactoryController implements IFactoryController
 {
 
     ////// --------------------------- NBTConstants --------------------------- \\\\\\
-    private static final String NBT_TYPE = "Type";
-    private static final String NBT_DATA = "Data";
+    static final String NBT_TYPE = "Type";
+    static final String NBT_DATA = "Data";
     ////// --------------------------- NBTConstants --------------------------- \\\\\\
 
     /**
      * Instance variable.
      */
-    private static final StandardFactoryController INSTANCE       = new StandardFactoryController();
+    private static final StandardFactoryController                    INSTANCE                = new StandardFactoryController();
+    /**
+     * Primary (main) Input mappings.
+     */
+    @NotNull
+    private final        HashMap<TypeToken, IFactory>                 primaryInputMappings    = new HashMap<>();
     /**
-     * Input mappings.
+     * Primary (main) Output mappings.
      */
-    @SuppressWarnings(RAWTYPES)
     @NotNull
-    private final        Map<Class, IFactory>      inputMappings  = new HashMap<>();
+    private final        HashMap<TypeToken, IFactory>                 primaryOutputMappings   = new HashMap<>();
     /**
-     * Output mappings.
+     * Secondary (super) Input mappings
      */
-    @SuppressWarnings(RAWTYPES)
     @NotNull
-    private final        Map<Class, IFactory>      outputMappings = new HashMap<>();
+    private final        HashMap<TypeToken, Set<IFactory>>            secondaryInputMappings  = new HashMap<>();
```

Comment:
The formatting seems quite strange here, did we agree once that we want if divided by javadoc default formatting?
Anyway it would be nice to know what the difference between a secondary and primary input mapping is.